### PR TITLE
Implemented Binary search to gain O(log n) lookup performance in find_region()

### PR DESCRIPTION
**Pull request by eminghuliev**
I have implemented Binary search in find_region which mentioned in #41 issue. There is UnsortedMemoryRegions mechanism to make sure sorting of memory regions during creating a new `GuestMemoryMmap` vector. That's why I just implemented binary search through sorted region map.

Signed-off-by: drmint80@gmail.com

### CODE REVIEW

**Comment by jiangliu** [comment]
I still have a question here. For most cases, we should have one or two memory ranges in an GuestMemory object, and the original search algorithm should be good enough. So what scenarios will have a relative more ranges in an GuestMemory object?

**Comment by jiangliu** [comment]
Or any performance comparison with the previous linear search algorithm? Does it achieve good performance with 1-2 memory ranges only?

**Comment by drmint80** [comment]
That's right it will give kinda performance when small range memory (even O(log n) in worst-case). As I know this project can be used for other purposes related to virtualization. If is there huge memory range for any other purposes, linear iteration can cause performance degradation.

**Comment by jiangliu** [comment]
Thanks for the clarification:)

File: src/mmap.rs [hunk 1]
```diff
@@ -447,14 +447,23 @@ impl GuestMemory for GuestMemoryMmap {
...
-    fn find_region(&self, addr: GuestAddress) -> Option<&GuestRegionMmap> {
-        for region in self.regions.iter() {
-            if addr >= region.start_addr() && addr <= region.end_addr() {
-                return Some(region);
-            }
-        }
-        None
-    }
+    fn find_region(&self, addr: GuestAddress) -> Option<&GuestRegionMmap> {           
+        let bs_region = self.regions.binary_search_by_key(&addr, |x| x.start_addr());
```
Comment (andreeaflorescu) [inline src/mmap.rs]: This looks like it's changing the behavior of `find_region`. The function was previously returning the region that contains the `addr` passed as a parameter. Now it looks like it returns the region that starts with `addr`. Is this true?

Comment (drmint80) [reply src/mmap.rs]: Return value is same with previous version. In previous version it just does iterate through regions. Now it returns just region with index (e.g regions[x]) which founded with BS.

Comment (drmint80) [reply src/mmap.rs]: It does same thing still. In BS version, it searches exact match (starts with `addr`) if can't find exact value that starts with `addr` then searches nearest value (check `end_addr` in like previous version) in line 458.

Comment (andreeaflorescu) [reply src/mmap.rs]: Ok, thanks. Let me check that again.

**Comment by bonzini** [comment]
Closing to reopen and include a rebased version.

**Pull request closed without merging**